File: sdks/python/agentsdk/backup.py
"""Backup — timestamped zip archives with atomic write and collision safety.

Provides ``CreateBackup`` to generate collision-safe, timestamped zip archives
of specified items, ``ListBackups`` to scan an output directory for existing
backup files sorted newest-first, ``GFSRotate`` for grandfather-father-son
retention rotation, and ``BackupConfig`` for load/save of backup configuration.

Filename format: ``{prefix}-backup-YYYYMMDD-HHMMSS.zip``
On collision, appends incrementing suffix: ``-2``, ``-3``, ...

Zip entry paths use forward slashes for cross-platform compatibility.
Non-existent items in the items list are silently skipped.
The zip is written atomically via a ``.tmp`` file + ``os.replace()``.
"""

from __future__ import annotations

import glob
import json
import logging
import os
import zipfile
from dataclasses import asdict, dataclass
from datetime import datetime
from pathlib import Path, PurePosixPath
from typing import Any, Dict, List, Optional, Tuple

logger = logging.getLogger(__name__)


@dataclass
class BackupMeta:
    """Metadata about a backup archive."""

    filename: str  # Base filename, e.g. "myapp-backup-20250101-120000.zip"
    size: int  # File size in bytes
    created_at: Optional[datetime]  # Parsed from filename timestamp


def _parse_backup_filename(filename: str, prefix: str) -> Optional[datetime]:
    """Extract the timestamp from a backup filename.

    Formats: ``{prefix}-backup-YYYYMMDD-HHMMSS.zip``
             ``{prefix}-backup-YYYYMMDD-HHMMSS-N.zip``
    """
    expected_prefix = f"{prefix}-backup-"
    if not filename.startswith(expected_prefix):
        return None

    rest = filename[len(expected_prefix):]
    # Remove .zip extension.
    rest = rest.removesuffix(".zip")

    # Check for collision suffix: YYYYMMDD-HHMMSS-N
    ts_part = rest
    # "YYYYMMDD-HHMMSS" is 15 chars; any dash at index 15+ indicates a suffix.
    last_dash = rest.rfind("-")
    if last_dash >= 15:
        suffix_str = rest[last_dash + 1:]
        if suffix_str.isdigit():
            ts_part = rest[:last_dash]

    try:
        return datetime.strptime(ts_part, "%Y%m%d-%H%M%S")
    except ValueError:
        return None


def CreateBackup(
    data_dir: str, output_dir: str, prefix: str, items: List[str]
) -> Tuple[str, int]:
    """Generate a collision-safe, timestamped zip archive.

    Args:
        data_dir: Root directory containing the items to back up.
        output_dir: Directory where the zip archive will be written.
        prefix: Filename prefix (e.g. ``"myapp"``).
        items: List of file or directory names relative to *data_dir*.

    Returns:
        A ``(zip_path, size)`` tuple.

    Raises:
        OSError: If *data_dir* does not exist or the zip cannot be written.
        ValueError: If *data_dir* is not a directory.
    """
    # Validate data_dir exists.
    if not os.path.exists(data_dir):
        raise OSError(f"backup: stat data dir {data_dir!r}: No such file or directory")
    if not os.path.isdir(data_dir):
        raise ValueError(f"backup: data dir {data_dir!r} is not a directory")

    # Create output directory if needed.
    try:
        os.makedirs(output_dir, exist_ok=True)
    except OSError as e:
        raise OSError(f"backup: create output dir {output_dir!r}: {e}") from e

    # Generate base filename from current timestamp.
    now = datetime.now()
    ts_str = now.strftime("%Y%m%d-%H%M%S")
    base_name = f"{prefix}-backup-{ts_str}.zip"

    # Resolve collision by appending -2, -3, ...
    target = os.path.join(output_dir, base_name)
    final_target = target
    suffix = 2
    while os.path.exists(final_target):
        final_target = os.path.join(
            output_dir, f"{prefix}-backup-{ts_str}-{suffix}.zip"
        )
        suffix += 1

    # Atomic write: create .tmp file, then rename.
    tmp_path = final_target + ".tmp"

    try:
        with zipfile.ZipFile(tmp_path, "w", compression=zipfile.ZIP_DEFLATED) as zw:
            for item in items:
                src_path = os.path.join(data_dir, item)
                if not os.path.exists(src_path):
                    # Silently skip non-existent items.
                    continue

                if os.path.isdir(src_path):
                    # Walk the directory tree.
                    for dirpath, dirnames, filenames in os.walk(src_path):
                        for fname in filenames:
                            full_path = os.path.join(dirpath, fname)
                            if not os.path.isfile(full_path):
                                # Skip non-regular files (symlinks, etc.)
                                continue
                            rel = os.path.relpath(full_path, data_dir)
                            zip_name = str(PurePosixPath(rel))
                            try:
                                zw.write(full_path, zip_name)
                            except OSError as e:
                                raise OSError(
                                    f"backup: add file {full_path!r}: {e}"
                                ) from e
                elif os.path.isfile(src_path):
                    zip_name = str(PurePosixPath(item))
                    try:
                        zw.write(src_path, zip_name)
                    except OSError as e:
                        raise OSError(
                            f"backup: add file {src_path!r}: {e}"
                        ) from e
                # Skip symlinks and other non-file, non-dir items silently.
    except Exception:
        # Clean up temp file on failure.
        try:
            os.remove(tmp_path)
        except OSError:
            pass
        raise

    # Atomic rename from .tmp to final target.
    try:
        os.replace(tmp_path, final_target)
    except OSError as e:
        try:
            os.remove(tmp_path)
        except OSError:
            pass
        raise OSError(
            f"backup: rename {tmp_path!r} to {final_target!r}: {e}"
        ) from e

    # Read back size.
    final_info = os.stat(final_target)
    return final_target, final_info.st_size


def ListBackups(output_dir: str, prefix: str) -> List[BackupMeta]:
    """Scan *output_dir* for backup files matching *prefix*.

    Returns metadata sorted by ``created_at`` descending (newest first).
    Returns an empty list if no backups are found or the directory does not
    exist.

    Args:
        output_dir: Directory containing backup zip files.
        prefix: Filename prefix to filter on.

    Returns:
        A list of :class:`BackupMeta`, always non-None (empty list if none found).
    """
    metas: List[BackupMeta] = []

    pattern = os.path.join(output_dir, f"{prefix}-backup-*.zip")
    matches = glob.glob(pattern)

    for m in matches:
        try:
            info = os.stat(m)
        except OSError:
            continue

        meta = BackupMeta(
            filename=os.path.basename(m),
            size=info.st_size,
            created_at=_parse_backup_filename(os.path.basename(m), prefix),
        )
        metas.append(meta)

    # Sort by created_at descending (newest first).
    # Items with None created_at go to the end.
    metas.sort(
        key=lambda b: b.created_at or datetime.min,
        reverse=True,
    )

    return metas


# ---------------------------------------------------------------------------
# GFS (Grandfather-Father-Son) rotation
# ---------------------------------------------------------------------------


@dataclass
class RetentionPolicy:
    """Defines how many backups to retain per time granularity."""

    daily: int = 0   # Number of newest backups to keep per calendar day.
    weekly: int = 0  # Number of newest backups to keep per ISO week.
    monthly: int = 0  # Number of newest backups to keep per calendar month.


@dataclass
class RotationResult:
    """Reports which backups were kept and which were removed by GFSRotate."""

    kept: List[str]  # Filenames of protected (kept) backups.
    removed: List[str]  # Filenames of successfully deleted backups.


def GFSRotate(
    backups: List[BackupMeta], policy: RetentionPolicy, output_dir: str
) -> RotationResult:
    """Perform grandfather-father-son rotation on the given backups.

    Backups are expected to be sorted newest-first (as returned by
    ``ListBackups``). For each retention rule (daily, weekly, monthly), the N
    newest backups within each group (calendar day, ISO week, calendar month)
    are marked as protected. A backup protected by ANY rule is kept.

    Unprotected backups are deleted from *output_dir*. Individual deletion
    failures are logged but do not halt rotation — the failed file remains in
    the Kept list.

    Args:
        backups: List of :class:`BackupMeta` sorted newest-first.
        policy: Retention policy defining daily/weekly/monthly limits.
        output_dir: Directory containing the backup files.

    Returns:
        A :class:`RotationResult` with kept and removed filenames.

    Raises:
        OSError: If *output_dir* does not exist or is not a directory.
    """
    # Validate output_dir.
    if not os.path.exists(output_dir):
        raise OSError(
            f"backup: stat output dir {output_dir!r}: "
            "No such file or directory"
        )
    if not os.path.isdir(output_dir):
        raise OSError(
            f"backup: output dir {output_dir!r} is not a directory"
        )

    result = RotationResult(kept=[], removed=[])

    if not backups:
        return result

    # Build protected set: filename → True.
    protected: Dict[str, bool] = {}

    # Group keys and counters.
    # group_key = (rule, group_string) → count of protected backups so far.
    group_counts: Dict[Tuple[str, str], int] = {}

    for b in backups:
        if b.created_at is None:
            continue

        iso_cal = b.created_at.isocalendar()
        iso_year = iso_cal[0]
        iso_week = iso_cal[1]
        month = b.created_at.strftime("%Y%m")
        day_key = b.created_at.strftime("%Y%m%d")
        week_key = f"{iso_year:04d}-W{iso_week:02d}"

        # Check each rule; a backup is protected if ANY rule protects it.
        if policy.daily > 0:
            dk = ("daily", day_key)
            if group_counts.get(dk, 0) < policy.daily:
                protected[b.filename] = True
                group_counts[dk] = group_counts.get(dk, 0) + 1

        if policy.weekly > 0:
            wk = ("weekly", week_key)
            if group_counts.get(wk, 0) < policy.weekly:
                protected[b.filename] = True
                group_counts[wk] = group_counts.get(wk, 0) + 1

        if policy.monthly > 0:
            mk = ("monthly", month)
            if group_counts.get(mk, 0) < policy.monthly:
                protected[b.filename] = True
                group_counts[mk] = group_counts.get(mk, 0) + 1

    # Separate kept and unprotected backups.
    unprotected: List[BackupMeta] = []
    for b in backups:
        if protected.get(b.filename):
            result.kept.append(b.filename)
        else:
            unprotected.append(b)

    # Delete unprotected backups from output_dir.
    for b in unprotected:
        path = os.path.join(output_dir, b.filename)
        try:
            os.remove(path)
            result.removed.append(b.filename)
        except OSError:
            # Log but don't abort — keep the file in Kept since deletion failed.
            logger.warning("backup: delete %r: failed", path)
            result.kept.append(b.filename)

    # Sort for deterministic output.
    result.kept.sort()
    result.removed.sort()

    return result


# ---------------------------------------------------------------------------
# Backup configuration
# ---------------------------------------------------------------------------


@dataclass
class BackupConfig:
    """Full backup configuration including retention policy and output directory."""

    retention_policy: RetentionPolicy
    output_dir: str


def DefaultBackupConfig() -> BackupConfig:
    """Return a :class:`BackupConfig` with sensible defaults.

    Daily=7, Weekly=4, Monthly=6, and empty OutputDir.
    """
    return BackupConfig(
        retention_policy=RetentionPolicy(daily=7, weekly=4, monthly=6),
        output_dir="",
    )


def LoadBackupConfig(file_path: str) -> BackupConfig:
    """Read a :class:`BackupConfig` from *file_path*.

    If the file does not exist, returns the default config.
    On JSON parse error, raises an OSError with ``backup: parse config`` prefix.

    Args:
        file_path: Path to the JSON configuration file.

    Returns:
        A :class:`BackupConfig`.

    Raises:
        OSError: On read or parse errors (except missing file → defaults).
    """
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            data = json.load(f)
    except FileNotFoundError:
        return DefaultBackupConfig()
    except OSError as e:
        raise OSError(f"backup: read config {file_path!r}: {e}") from e
    except json.JSONDecodeError as e:
        raise OSError(f"backup: parse config: {e}") from e

    # Map JSON keys to dataclass fields.
    rp_data = data.get("retention_policy", {})
    retention = RetentionPolicy(
        daily=rp_data.get("daily", 7),
        weekly=rp_data.get("weekly", 4),
        monthly=rp_data.get("monthly", 6),
    )

    return BackupConfig(
        retention_policy=retention,
        output_dir=data.get("output_dir", ""),
    )


def SaveBackupConfig(file_path: str, config: BackupConfig) -> None:
    """Write *config* to *file_path* as indented JSON.

    Creates parent directories as needed and uses atomic write
    (write to ``.tmp``, then ``os.replace``). Any pre-existing ``.tmp``
    file from a prior crash is overwritten.

    Args:
        file_path: Destination path for the JSON file.
        config: The :class:`BackupConfig` to serialize.

    Raises:
        OSError: On write, directory creation, or rename failure.
    """
    # Build the JSON-serializable dict matching Go SDK's field names.
    obj: Dict[str, Any] = {
        "retention_policy": {
            "daily": config.retention_policy.daily,
            "weekly": config.retention_policy.weekly,
            "monthly": config.retention_policy.monthly,
        },
        "output_dir": config.output_dir,
    }

    try:
        data = json.dumps(obj, indent=2)
    except (TypeError, ValueError) as e:
        raise OSError(f"backup: marshal config: {e}") from e

    # Create parent directories.
    parent = os.path.dirname(file_path)
    try:
        os.makedirs(parent, exist_ok=True)
    except OSError as e:
        raise OSError(f"backup: create config dir {parent!r}: {e}") from e

    # Atomic write: .tmp + os.replace.
    tmp_file = file_path + ".tmp"
    try:
        with open(tmp_file, "w", encoding="utf-8") as f:
            f.write(data)
    except OSError as e:
        raise OSError(f"backup: write config tmp {tmp_file!r}: {e}") from e

    try:
        os.replace(tmp_file, file_path)
    except OSError as e:
        try:
            os.remove(tmp_file)
        except OSError:
            pass
        raise OSError(
            f"backup: rename config {tmp_file!r} → {file_path!r}: {e}"
        ) from e

File: sdks/go/backup.go
package agentsdk

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// BackupMeta holds metadata about a backup archive.
type BackupMeta struct {
	Filename  string    // Base filename, e.g. "myapp-backup-20250101-120000.zip"
	Size      int64     // File size in bytes
	CreatedAt time.Time // Parsed from filename timestamp
}

// CreateBackup generates a collision-safe, timestamped zip archive of the
// specified items relative to dataDir and writes it to outputDir.
//
// Filename format: {prefix}-backup-YYYYMMDD-HHMMSS.zip
// On collision, appends incrementing suffix: -2, -3, ...
//
// Zip entry paths use forward slashes for cross-platform compatibility.
// Non-existent items in the items slice are silently skipped.
// The zip is written atomically via a .tmp file + os.Rename.
func CreateBackup(dataDir, outputDir, prefix string, items []string) (zipPath string, size int64, err error) {
	// Validate dataDir exists.
	info, err := os.Stat(dataDir)
	if err != nil {
		return "", 0, fmt.Errorf("backup: stat data dir %q: %w", dataDir, err)
	}
	if !info.IsDir() {
		return "", 0, fmt.Errorf("backup: data dir %q is not a directory", dataDir)
	}

	// Create output directory if needed.
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return "", 0, fmt.Errorf("backup: create output dir %q: %w", outputDir, err)
	}

	// Generate base filename from current timestamp.
	now := time.Now()
	baseName := fmt.Sprintf("%s-backup-%s.zip", prefix, now.Format("20060102-150405"))

	// Resolve collision by appending -2, -3, ...
	target := filepath.Join(outputDir, baseName)
	finalTarget := target
	for i := 2; ; i++ {
		if _, err := os.Stat(finalTarget); os.IsNotExist(err) {
			break
		}
		if err != nil {
			return "", 0, fmt.Errorf("backup: check existing file %q: %w", finalTarget, err)
		}
		finalTarget = filepath.Join(outputDir, fmt.Sprintf("%s-backup-%s-%d.zip", prefix, now.Format("20060102-150405"), i))
	}

	// Atomic write: create .tmp file, then rename.
	tmpPath := finalTarget + ".tmp"

	var writeErr error
	func() {
		f, err := os.Create(tmpPath)
		if err != nil {
			writeErr = fmt.Errorf("backup: create temp file %q: %w", tmpPath, err)
			return
		}
		defer func() {
			if cerr := f.Close(); cerr != nil && writeErr == nil {
				writeErr = fmt.Errorf("backup: close temp file %q: %w", tmpPath, cerr)
			}
		}()

		zw := zip.NewWriter(f)
		defer func() {
			if cerr := zw.Close(); cerr != nil && writeErr == nil {
				writeErr = fmt.Errorf("backup: close zip writer: %w", cerr)
			}
		}()

		for _, item := range items {
			srcPath := filepath.Join(dataDir, item)
			itemInfo, err := os.Stat(srcPath)
			if err != nil {
				// Silently skip non-existent or inaccessible items.
				continue
			}

			if itemInfo.IsDir() {
				// Walk the directory tree.
				err = filepath.Walk(srcPath, func(walkPath string, walkInfo os.FileInfo, walkErr error) error {
					if walkErr != nil {
						return walkErr
					}
					if walkInfo.IsDir() {
						return nil
					}

					rel, err := filepath.Rel(dataDir, walkPath)
					if err != nil {
						return fmt.Errorf("backup: compute relative path for %q: %w", walkPath, err)
					}
					// Force forward slashes in zip entry names.
					zipName := filepath.ToSlash(rel)

					if err := addFileToZip(zw, walkPath, zipName); err != nil {
						return fmt.Errorf("backup: add file %q: %w", walkPath, err)
					}
					return nil
				})
				if err != nil {
					writeErr = fmt.Errorf("backup: walk directory %q: %w", srcPath, err)
					return
				}
			} else {
				// Single file.
				zipName := filepath.ToSlash(item)
				if err := addFileToZip(zw, srcPath, zipName); err != nil {
					writeErr = fmt.Errorf("backup: add file %q: %w", srcPath, err)
					return
				}
			}
		}
	}()

	if writeErr != nil {
		// Clean up temp file on failure.
		os.Remove(tmpPath)
		return "", 0, writeErr
	}

	// Atomic rename from .tmp to final target.
	if err := os.Rename(tmpPath, finalTarget); err != nil {
		os.Remove(tmpPath)
		return "", 0, fmt.Errorf("backup: rename %q to %q: %w", tmpPath, finalTarget, err)
	}

	// Read back size.
	finalInfo, err := os.Stat(finalTarget)
	if err != nil {
		return finalTarget, 0, fmt.Errorf("backup: stat output file %q: %w", finalTarget, err)
	}

	return finalTarget, finalInfo.Size(), nil
}

// addFileToZip writes a single file to a zip writer.
func addFileToZip(zw *zip.Writer, filePath, zipName string) error {
	f, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("open %q: %w", filePath, err)
	}
	defer f.Close()

	info, err := f.Stat()
	if err != nil {
		return fmt.Errorf("stat %q: %w", filePath, err)
	}

	hdr, err := zip.FileInfoHeader(info)
	if err != nil {
		return fmt.Errorf("file info header for %q: %w", filePath, err)
	}
	hdr.Name = zipName
	// Use zero compression method for deterministic sizes (optional, but simplifies testing).
	// Standard deflate is fine for production; use stored for test stability.
	hdr.Method = zip.Deflate

	w, err := zw.CreateHeader(hdr)
	if err != nil {
		return fmt.Errorf("create zip entry %q: %w", zipName, err)
	}

	if _, err := io.Copy(w, f); err != nil {
		return fmt.Errorf("write zip entry %q: %w", zipName, err)
	}

	return nil
}

// ListBackups scans outputDir for files matching {prefix}-backup-*.zip and
// returns their metadata sorted by CreatedAt descending (newest first).
// Returns an empty (non-nil) slice if no backups are found or the directory
// does not exist.
func ListBackups(outputDir, prefix string) ([]BackupMeta, error) {
	metas := make([]BackupMeta, 0) // Always non-nil empty slice.

	pattern := prefix + "-backup-*.zip"
	matches, err := filepath.Glob(filepath.Join(outputDir, pattern))
	if err != nil {
		return nil, fmt.Errorf("backup: glob %q: %w", pattern, err)
	}

	for _, m := range matches {
		info, err := os.Stat(m)
		if err != nil {
			// Skip files we can't stat.
			continue
		}

		meta := BackupMeta{
			Filename: filepath.Base(m),
			Size:     info.Size(),
		}

		// Parse timestamp from filename.
		// Expected: {prefix}-backup-YYYYMMDD-HHMMSS.zip or {prefix}-backup-YYYYMMDD-HHMMSS-N.zip
		meta.CreatedAt, err = parseBackupFilename(meta.Filename, prefix)
		if err != nil {
			// If we can't parse the timestamp, use zero time.
			meta.CreatedAt = time.Time{}
		}

		metas = append(metas, meta)
	}

	// Sort by CreatedAt descending (newest first).
	// Files with zero time go to the end.
	for i := 0; i < len(metas)-1; i++ {
		for j := i + 1; j < len(metas); j++ {
			if metas[j].CreatedAt.After(metas[i].CreatedAt) {
				metas[i], metas[j] = metas[j], metas[i]
			}
		}
	}

	return metas, nil
}

// parseBackupFilename extracts the timestamp from a backup filename.
// Formats: {prefix}-backup-YYYYMMDD-HHMMSS.zip or {prefix}-backup-YYYYMMDD-HHMMSS-N.zip
func parseBackupFilename(filename, prefix string) (time.Time, error) {
	expectedPrefix := prefix + "-backup-"
	if !strings.HasPrefix(filename, expectedPrefix) {
		return time.Time{}, fmt.Errorf("backup: filename %q does not have prefix %q", filename, expectedPrefix)
	}

	rest := filename[len(expectedPrefix):]
	// Remove .zip extension.
	rest = strings.TrimSuffix(rest, ".zip")

	// Check for collision suffix: YYYYMMDD-HHMMSS-N
	tsPart := rest
	// "YYYYMMDD-HHMMSS" is 15 chars, dash at index 8 is part of timestamp.
	// Any dash at index 15 or later indicates a collision suffix (-N).
	lastDash := strings.LastIndex(rest, "-")
	if lastDash >= 15 {
		suffixStr := rest[lastDash+1:]
		if _, err := fmt.Sscanf(suffixStr, "%d", new(int)); err == nil {
			tsPart = rest[:lastDash]
		}
	}

	// Parse "YYYYMMDD-HHMMSS"
	return time.ParseInLocation("20060102-150405", tsPart, time.Local)
}

// RetentionPolicy defines how many backups to retain per time granularity.
type RetentionPolicy struct {
	Daily   int `json:"daily"`   // Number of newest backups to keep per calendar day.
	Weekly  int `json:"weekly"`  // Number of newest backups to keep per ISO week.
	Monthly int `json:"monthly"` // Number of newest backups to keep per calendar month.
}

// RotationResult reports which backups were kept and which were removed by GFSRotate.
type RotationResult struct {
	Kept    []string // Filenames of protected (kept) backups.
	Removed []string // Filenames of successfully deleted backups.
}

// GFSRotate performs grandfather-father-son rotation on the given backups.
//
// Backups are already expected to be sorted newest-first (as returned by
// ListBackups). For each retention rule (daily, weekly, monthly), the N newest
// backups within each group (calendar day, ISO week, calendar month) are marked
// as protected. A backup protected by ANY rule is kept.
//
// Unprotected backups are deleted from outputDir. Individual deletion failures
// are logged but do not halt rotation — the failed file remains in the Kept list.
//
// Returns an error only if outputDir does not exist or is not a directory.
func GFSRotate(backups []BackupMeta, policy RetentionPolicy, outputDir string) (RotationResult, error) {
	// Validate outputDir.
	info, err := os.Stat(outputDir)
	if err != nil {
		return RotationResult{}, fmt.Errorf("backup: stat output dir %q: %w", outputDir, err)
	}
	if !info.IsDir() {
		return RotationResult{}, fmt.Errorf("backup: output dir %q is not a directory", outputDir)
	}

	result := RotationResult{
		Kept:    make([]string, 0),
		Removed: make([]string, 0),
	}

	if len(backups) == 0 {
		return result, nil
	}

	// Build protected set: filename → true.
	protected := make(map[string]bool)

	// Group keys and counters.
	type groupKey struct {
		rule  string // "daily", "weekly", "monthly"
		group string // YYYYMMDD, YYYY-WNN, or YYYYMM
	}
	groupCounts := make(map[groupKey]int)

	for _, b := range backups {
		isoYear, week := b.CreatedAt.ISOWeek()
		month := b.CreatedAt.Format("200601")
		dayKey := b.CreatedAt.Format("20060102")
		weekKey := fmt.Sprintf("%04d-W%02d", isoYear, week)

		// Check each rule; a backup is protected if ANY rule protects it.
		dailyKey := groupKey{"daily", dayKey}
		weeklyKey := groupKey{"weekly", weekKey}
		monthlyKey := groupKey{"monthly", month}

		if policy.Daily > 0 && groupCounts[dailyKey] < policy.Daily {
			protected[b.Filename] = true
			groupCounts[dailyKey]++
		}
		if policy.Weekly > 0 && groupCounts[weeklyKey] < policy.Weekly {
			protected[b.Filename] = true
			groupCounts[weeklyKey]++
		}
		if policy.Monthly > 0 && groupCounts[monthlyKey] < policy.Monthly {
			protected[b.Filename] = true
			groupCounts[monthlyKey]++
		}
	}

	// Separate kept and unprotected backups.
	var unprotected []BackupMeta
	for _, b := range backups {
		if protected[b.Filename] {
			result.Kept = append(result.Kept, b.Filename)
		} else {
			unprotected = append(unprotected, b)
		}
	}

	// Delete unprotected backups from outputDir.
	for _, b := range unprotected {
		path := filepath.Join(outputDir, b.Filename)
		if err := os.Remove(path); err != nil {
			// Log but don't abort — keep the file in Kept since deletion failed.
			log.Printf("backup: delete %q: %v", path, err)
			result.Kept = append(result.Kept, b.Filename)
		} else {
			result.Removed = append(result.Removed, b.Filename)
		}
	}

	// Sort for deterministic output.
	sort.Strings(result.Kept)
	sort.Strings(result.Removed)

	return result, nil
}

// BackupConfig holds the full backup configuration including retention policy
// and output directory.
type BackupConfig struct {
	RetentionPolicy RetentionPolicy `json:"retention_policy"`
	OutputDir       string          `json:"output_dir"`
}

// DefaultBackupConfig returns a BackupConfig with sensible defaults:
// Daily=7, Weekly=4, Monthly=6, and empty OutputDir.
func DefaultBackupConfig() BackupConfig {
	return BackupConfig{
		RetentionPolicy: RetentionPolicy{
			Daily:   7,
			Weekly:  4,
			Monthly: 6,
		},
		OutputDir: "",
	}
}

// LoadBackupConfig reads a BackupConfig from filePath. If the file does not
// exist, it returns the default config. On JSON parse error, the error is
// wrapped with "backup: parse config".
func LoadBackupConfig(filePath string) (BackupConfig, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return DefaultBackupConfig(), nil
		}
		return BackupConfig{}, fmt.Errorf("backup: read config %q: %w", filePath, err)
	}

	var cfg BackupConfig
	if err := json.Unmarshal(data, &cfg); err != nil {
		return BackupConfig{}, fmt.Errorf("backup: parse config: %w", err)
	}

	return cfg, nil
}

// SaveBackupConfig writes cfg to filePath as indented JSON. It creates parent
// directories as needed and uses atomic write (write to .tmp, then os.Rename).
// Any pre-existing .tmp file from a prior crash is overwritten.
func SaveBackupConfig(filePath string, cfg BackupConfig) error {
	data, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return fmt.Errorf("backup: marshal config: %w", err)
	}

	dir := filepath.Dir(filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("backup: create config dir %q: %w", dir, err)
	}

	tmpFile := filePath + ".tmp"
	if err := os.WriteFile(tmpFile, data, 0644); err != nil {
		return fmt.Errorf("backup: write config tmp %q: %w", tmpFile, err)
	}

	if err := os.Rename(tmpFile, filePath); err != nil {
		// Clean up the temp file on rename failure.
		os.Remove(tmpFile)
		return fmt.Errorf("backup: rename config %q → %q: %w", tmpFile, filePath, err)
	}

	return nil
}
